From ae5e1287b56664b486629f67c451248db34dd7d1 Mon Sep 17 00:00:00 2001
From: Anmol Jain <ajain@ciq.com>
Date: Wed, 30 Jul 2025 10:01:43 +0000
Subject: [PATCH] netfilter: nf_tables: skip bound chain on rule flush

jira VULN-6585
cve CVE-2023-3777
commit-author Pablo Neira Ayuso <pablo@netfilter.org>
commit 6eaf41e87a223ae6f8e7a28d6e78384ad7e407f8

Skip bound chain when flushing table rules, the rule that owns this
chain releases these objects.

Otherwise, the following warning is triggered:

  WARNING: CPU: 2 PID: 1217 at net/netfilter/nf_tables_api.c:2013 nf_tables_chain_destroy+0x1f7/0x210 [nf_tables]
  CPU: 2 PID: 1217 Comm: chain-flush Not tainted 6.1.39 #1
  RIP: 0010:nf_tables_chain_destroy+0x1f7/0x210 [nf_tables]

Fixes: d0e2c7de92c7 ("netfilter: nf_tables: add NFT_CHAIN_BINDING")
	Reported-by: Kevin Rich <kevinrich1337@gmail.com>
	Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
	Signed-off-by: Florian Westphal <fw@strlen.de>
(cherry picked from commit 6eaf41e87a223ae6f8e7a28d6e78384ad7e407f8)
	Signed-off-by: Anmol Jain <ajain@ciq.com>
---
 net/netfilter/nf_tables_api.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 88dd8669336d..5264437d3b89 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3756,6 +3756,8 @@ static int nf_tables_delrule(struct sk_buff *skb, const struct nfnl_info *info,
 		list_for_each_entry(chain, &table->chains, list) {
 			if (!nft_is_active_next(net, chain))
 				continue;
+			if (nft_chain_is_bound(chain))
+				continue;
 
 			ctx.chain = chain;
 			err = nft_delrule_by_chain(&ctx);
-- 
2.47.3

